mllib/src/main/scala/org/apache/spark/mllib/linalg/BLAS.scala

```diff
@@ -473,13 +473,16 @@ private[spark] object BLAS extends Serializable with Logging {
     if (alpha == 0.0) {
       logDebug("gemv: alpha is equal to 0. Returning y.")
     } else {
-      A match {
-        case sparse: SparseMatrix =>
-          gemv(alpha, sparse, x, beta, y)
-        case dense: DenseMatrix =>
-          gemv(alpha, dense, x, beta, y)
+      (A, x) match {
+        case (sparse: SparseMatrix, dx: DenseVector) =>
+          gemv(alpha, sparse, dx, beta, y)
+        case (dense: DenseMatrix, dx: DenseVector) =>
+          gemv(alpha, dense, dx, beta, y)
+        case (dense: DenseMatrix, sx: SparseVector) =>
+          gemv(alpha, dense, sx, beta, y)
         case _ =>
```

Comment:
How about `SparseMatrix` and `SparseVector`? To make the consistent naming, we can use `dmA`, `smA`, `dvx`, and `svx`.